### PR TITLE
[DML EP] Add RotaryEmbedding

### PR DESCRIPTION
**Pull request by PatriceVignola**
This is a graph implementation of RotaryEmbedding since there's no time to add it to DML before 1.16.2, but it eventually should move into DirectML since we're bandwidth-bound.

### CODE REVIEW

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+
+        ML_CHECK_VALID_ARGUMENT(m_inputTensorDescs[inputDataIndex].GetDimensionCount() == 4);
+        ML_CHECK_VALID_ARGUMENT(m_inputTensorDescs[positionIdsIndex].GetDimensionCount() == 4);
+        ML_CHECK_VALID_ARGUMENT(m_inputTensorDescs[cosCacheIndex].GetDimensionCount() == 4);
+        ML_CHECK_VALID_ARGUMENT(m_inputTensorDescs[sinCacheIndex].GetDimensionCount() == 4);
+
+        ML_CHECK_VALID_ARGUMENT(m_outputTensorDescs[0].GetDimensionCount() == 4);
+
+        ML_CHECK_VALID_ARGUMENT(m_inputTensorDescs[cosCacheIndex].GetSizes() == m_inputTensorDescs[sinCacheIndex].GetSizes());
+        const uint32_t headSize = m_inputTensorDescs[cosCacheIndex].GetSizes().back() * 2;
```
Comment (sumitsays) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp:59]: If I am not mistaken, gsl::span == operator compares the pointer not the actual values.
[https://github.com/microsoft/GSL/blob/e64c97fc2cfc11992098bb38eda932de275e3f4d/include/gsl/span#L247](https://github.com/microsoft/GSL/blob/e64c97fc2cfc11992098bb38eda932de275e3f4d/include/gsl/span#L247)

If that's what intended here, then I think `cosCache` and `sinCache` will not have same gsl::span object.

Comment (PatriceVignola) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp:59]: You're looking at the `span_iterator` comparison. The `span` comparison is over here: https://github.com/microsoft/GSL/blob/e64c97fc2cfc11992098bb38eda932de275e3f4d/include/gsl/span_ext#L57C8-L57C8

Comment (sumitsays) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp:59]: Ahh I see.

**Review by sumitsays — DISMISSED** [review]

**Review by sumitsays — APPROVED** [review]

**Pull request merged**